### PR TITLE
copr: added sub duration and string

### LINKED ISSUE #9016: Coprocessor functions migration from non-vec framework
## Feature Request

Coprocessor historically has two implementations. One is based on the vectorized framework, and the other one is based on the plain (non-vec) framework, which was removed recently. A few of the functions in the legacy non-vec framework have not been ported into the vectorized framework yet.

The following is the list of functions that existed in non-vec framework but not in vectorized framework. You may also want to look into the [non-vec framework before it was removed](https://github.com/tikv/tikv/tree/1a88f12ebf50064b992fc5efc7e7f55795210521/components/tidb_query_normal_expr/src).

- [x] AddDateAndDuration
- [ ] AddDateAndString
- [x] AddDatetimeAndDuration
- [x] AddDatetimeAndString
- [x] AddDurationAndDuration
- [x] AddDurationAndString
- [ ] AddTimeDateTimeNull
- [ ] AddTimeDurationNull
- [ ] AddTimeStringNull
- [ ] Compress
- [x] Date
- [ ] DateDiff
- [ ] Instr
- [x] Locate2ArgsUtf8
- [x] Locate3ArgsUtf8
- [ ] Lower
- [x] NullTimeDiff
- [x] Quote
- [x] RegexpSig
- [x] RegexpUtf8Sig
- [ ] RpadUtf8
- [x] SubDateAndDuration
- [ ] SubDateAndString
- [x] SubDatetimeAndDuration
- [ ] SubDatetimeAndString
- [x] SubDurationAndDuration
- [x] SubDurationAndString
- [ ] SubTimeDateTimeNull
- [ ] SubTimeDurationNull
- [ ] Substring2Args
- [ ] Substring2ArgsUtf8
- [ ] Substring3Args
- [ ] Substring3ArgsUtf8
- [ ] Trim2Args
- [ ] TruncateDecimal
- [ ] TruncateUint
- [ ] Uncompress
- [x] WeekWithoutMode
- [x] YearWeekWithMode
- [x] YearWeekWithoutMode

### PR DESCRIPTION
**Pull request by aswin6197**
signed-off-by: aswin <aswin6197@gmail.com>
### What problem does this PR solve?

Issue Number: close #9016  

Problem Summary:

### What is changed and how it works?
What's Changed: Added vectorize sub duration and string

### Check List 
Tests 
- Unit test

### Release note 
- No release note

### CODE REVIEW

**Comment by andylokandy** [comment]
Please fix dco and tests

File: components/tidb_query_expr/src/impl_time.rs [hunk 1]
```diff
@@ -700,6 +700,33 @@ pub fn add_duration_and_string(
...
+        Err(_) => return Ok(None),
+    };
+
+    let res = match arg1.checked_sub(arg2) {
+        Some(res) => res,
+        None => {
+            return ctx
+                .handle_invalid_time_error(Error::overflow(
+                    "DURATION",
+                    format!("({} + {})", arg1, arg2),
```
Comment (andylokandy) [inline components/tidb_query_expr/src/impl_time.rs:722]:
```suggestion
                    format!("({} - {})", arg1, arg2),
```

**Comment by aswin6197** [comment]
I added the signature by git commit, not sure why DCO is still failing

File: components/tidb_query_expr/src/impl_time.rs (hunk 1 above)
Comment (aswin6197) [reply components/tidb_query_expr/src/impl_time.rs]: I changed the + to -
I added the git sign, but DCO is still failing, Is there something else I should add

Comment (aswin6197) [reply components/tidb_query_expr/src/impl_time.rs]: The unit test on the file i changed are passing but when i run the make test command, the make file exists with error code 101, running test on the commit before mine is also giving the same error.
So i'm not sure why it's failing

**Comment by aswin6197** [comment]
The unit test i wrote are passing when i run them directly but the  make test is failing on some other part.
I didn't modify any other file, so not sure why it's failing

**Comment by andylokandy** [comment]
You may try squash the commits and force push.

File: components/tidb_query_expr/src/impl_time.rs [hunk 2]
```diff
@@ -765,11 +792,36 @@ mod tests {
...
+            (Some("24:00:01"), Some("00:00:02"), Some("23:59:59")),
+            // (None, None, None),
+        ];
+        let mut ctx = EvalContext::default();
+        for (duration, string, exp) in cases {
+            let expected = exp.map(|exp| Duration::parse(&mut ctx, exp, MAX_FSP).unwrap());
+            let duration = duration.map(|arg1| Duration::parse(&mut ctx, arg1, MAX_FSP).unwrap());
+            let string = string.map(|arg2| arg2.as_bytes().to_vec());
+
+            // let string = string.map(|arg2| Duration::parse(&mut ctx, arg2, MAX_FSP).unwrap());
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_time.rs:820]: remove not useful comment.

File: components/tidb_query_expr/src/impl_time.rs [hunk 3]
```diff
@@ -765,11 +792,36 @@ mod tests {
...
+    }
+
+    #[test]
+    fn test_sub_duration_and_string() {
+        let cases = vec![
+            (Some("00:02:02"), Some("00:01:01"), Some("00:01:01")),
+            (Some("12:00:00"), Some("00:00:01"), Some("11:59:59")),
+            (Some("24:00:00"), Some("00:00:01"), Some("23:59:59")),
+            (Some("24:00:01"), Some("00:00:02"), Some("23:59:59")),
+            // (None, None, None),
```
Comment (Xuanwo) [inline components/tidb_query_expr/src/impl_time.rs:812]: Why remove this test?

**Comment by andylokandy** [comment]
@aswin6197 Thanks for your contribution. Would you like to fix/answer the comments? I'd like to merge it soon after that.

**Comment by aswin6197** [comment]
I made the changes Xuanwo suggested.
Sorry about the delay.

**Comment by andylokandy** [comment]
@aswin6197 Thank you for your patient, but the test still fails, would you like to take a look?

**Comment by aswin6197** [comment]
I think i might have messed up trying to fix the conflicts, should i just make a fork from the master and redo it.

**Pull request closed without merging**